src/Env.hs

```diff
@@ -0,0 +1,320 @@
+              then e {errors = (errors e) ++ ("Name already exsist in the context \n")}
+              else e {names = a : (names e)}
+
+addDeclaredName :: String -> VarEnv -> VarEnv
+addDeclaredName a e = if (a `elem` (declaredNames e))
+                      then e { errors = (errors e) ++ ("Name already exsist in the context \n") }
+                      else e { declaredNames = a : (declaredNames e) }
+
+--------------------------------------- Env Data Types ---------------------------------------
+
+-- | Environment for the dsll semantic checker. As the 'check' function executes, it
+-- accumulate information such as symbol tables in the environment.
+
+-- | list of elements that might appear in the global context
+
+data Ttype = Ttype { yt :: Y,
+                     kt :: K }
+             deriving (Show, Eq, Typeable)
+
+data TypeConstructor = TypeConstructor { nametc :: String,
+                                         klstc  :: [K],
+                                         typtc  :: Type }
+                       deriving (Show, Eq, Typeable)
+
+data VarConstructor = VarConstructor { namevc :: String,
+                                       ylsvc  :: [Y],
+                                       klsvc  :: [K],
+                                       tlsvc  :: [T],
+                                       tvc    :: T }
+                      deriving (Show, Eq, Typeable)
+
+data Operation = Operation { nameop :: String,
+                             ylsop  :: [Y],
+                             klsop  :: [K],
+                             tlsop  :: [T],
+                             top  :: T}
+                 deriving (Show, Eq, Typeable)
+
+data PredicateEnv = Pred1 Predicate1
+                  | Pred2 Predicate2
+                  deriving (Show, Eq, Typeable)
+
+data Predicate1 = Predicate1 { namepred1 :: String,
+                               ylspred1  :: [Y],
+                               klspred1  :: [K],
+                               tlspred1  :: [T],
+                               ppred1    :: Prop}
+                  deriving (Show, Eq, Typeable)
+
+data Predicate2 = Predicate2 { namepred2 :: String,
+                               plspred2  :: [Prop],
+                               ppred2    :: Prop }
+                  deriving (Show, Eq, Typeable)
+
+data VarEnv = VarEnv { typeConstructors :: M.Map String TypeConstructor,
+                       varConstructors  :: M.Map String VarConstructor,
+                       operations       :: M.Map String Operation,
+                       predicates       :: M.Map String PredicateEnv,
+                       typeVarMap       :: M.Map TypeVar Type,
+                       varMap           :: M.Map Var T,
+                       names            :: [String],  -- a global list which contains all the names of types in that env
+                       declaredNames    :: [String],  -- a global list which contains all the names of elements declared in that env
```

Comment:
rename to something like "varNames"? I'm not clear on what this is for. maybe this should be called "typeCtorNames"? or "dsllNames"?